Cloud storage plugin socket usage at capacity

### Link to reproduction

_No response_

### Describe the Bug

I am using the cloud storage plugin with an S3 adapter pointing to DigitalOcean Spaces. I have users making a lot of edits relating to images lately. After a significant amount of uploads and image swapping, I get reports of performance issues on the admin dashboard and uploads failing. I looked at the logs and I found these error messages:

```
@smithy/node-http-handler:WARN socket usage at capacity=50 and 102 additional requests are enqueued. See https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-maxsockets.html or increase socketAcquisitionWarningTimeout=(millis) in the NodeHttpHandler config.
@smithy/node-http-handler:WARN socket usage at capacity=50 and 114 additional requests are enqueued. See https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-maxsockets.html or increase socketAcquisitionWarningTimeout=(millis) in the NodeHttpHandler config.
@smithy/node-http-handler:WARN socket usage at capacity=50 and 120 additional requests are enqueued. See https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-maxsockets.html or increase socketAcquisitionWarningTimeout=(millis) in the NodeHttpHandler config.
@smithy/node-http-handler:WARN socket usage at capacity=50 and 121 additional requests are enqueued. See https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-maxsockets.html or increase socketAcquisitionWarningTimeout=(millis) in the NodeHttpHandler config.
@smithy/node-http-handler:WARN socket usage at capacity=50 and 125 additional requests are enqueued. See https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-maxsockets.html or increase socketAcquisitionWarningTimeout=(millis) in the NodeHttpHandler config.
@smithy/node-http-handler:WARN socket usage at capacity=50 and 135 additional requests are enqueued. See https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-maxsockets.html or increase socketAcquisitionWarningTimeout=(millis) in the NodeHttpHandler config.
@smithy/node-http-handler:WARN socket usage at capacity=50 and 145 additional requests are enqueued. See https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-maxsockets.html or increase socketAcquisitionWarningTimeout=(millis) in the NodeHttpHandler config.
@smithy/node-http-handler:WARN socket usage at capacity=50 and 156 additional requests are enqueued. See https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-maxsockets.html or increase socketAcquisitionWarningTimeout=(millis) in the NodeHttpHandler config.
@smithy/node-http-handler:WARN socket usage at capacity=50 and 158 additional requests are enqueued. See https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-maxsockets.html or increase socketAcquisitionWarningTimeout=(millis) in the NodeHttpHandler config.
@smithy/node-http-handler:WARN socket usage at capacity=50 and 161 additional requests are enqueued. See https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-maxsockets.html or increase socketAcquisitionWarningTimeout=(millis) in the NodeHttpHandler config.
@smithy/node-http-handler:WARN socket usage at capacity=50 and 163 additional requests are enqueued. See https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-maxsockets.html or increase socketAcquisitionWarningTimeout=(millis) in the NodeHttpHandler config.
@smithy/node-http-handler:WARN socket usage at capacity=50 and 166 additional requests are enqueued. See https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-maxsockets.html or increase socketAcquisitionWarningTimeout=(millis) in the NodeHttpHandler config.
@smithy/node-http-handler:WARN socket usage at capacity=50 and 169 additional requests are enqueued. See https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-maxsockets.html or increase socketAcquisitionWarningTimeout=(millis) in the NodeHttpHandler config.
@smithy/node-http-handler:WARN socket usage at capacity=50 and 170 additional requests are enqueued. See https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/node-configuring-maxsockets.html or increase socketAcquisitionWarningTimeout=(millis) in the NodeHttpHandler config.
```
These socket capacity warnings are sometimes preceded by error messages like this:

```
ERROR (payload): UnknownError
    err: {
      "type": "NoSuchKey",
      "message": "UnknownError",
      "stack":
          NoSuchKey: UnknownError
              at de_NoSuchKeyRes (/home/node/app/node_modules/@aws-sdk/client-s3/dist-cjs/index.js:4809:21)
              at de_CommandError (/home/node/app/node_modules/@aws-sdk/client-s3/dist-cjs/index.js:4747:19)
              at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
              at async /home/node/app/node_modules/@smithy/middleware-serde/dist-cjs/index.js:35:20
              at async /home/node/app/node_modules/@aws-sdk/middleware-signing/dist-cjs/index.js:225:18
              at async /home/node/app/node_modules/@smithy/middleware-retry/dist-cjs/index.js:320:38
              at async /home/node/app/node_modules/@aws-sdk/middleware-flexible-checksums/dist-cjs/index.js:173:18
              at async /home/node/app/node_modules/@aws-sdk/middleware-sdk-s3/dist-cjs/index.js:97:20
              at async /home/node/app/node_modules/@aws-sdk/middleware-sdk-s3/dist-cjs/index.js:120:14
              at async /home/node/app/node_modules/@aws-sdk/middleware-logger/dist-cjs/index.js:33:22
      "name": "NoSuchKey",
      "$fault": "client",
      "$metadata": {
        "httpStatusCode": 404,
        "requestId": "[request id]",
        "attempts": 1,
        "totalRetryDelay": 0
      },
      "Code": "NoSuchKey",
      "BucketName": "bucket-name",
      "RequestId": "[request id]",
      "HostId": "[host id]"
    }
```
I have my Payload app hosted on a basic DigitalOcean Droplet (1 vCPU, 1GB RAM, 25GB storage, added 1GB swap space) if that helps at all.

### To Reproduce

It seems like it happens when a lot of uploading is done. I haven't seen a clear-cut way to reproduce, but the best I can do is:

1. Set up a payload app with the cloud storage plugin
2. Point to a DigitalOcean Spaces bucket
3. Create an upload collection
4. Upload images in quick succession. Upload new images into existing records. Whatever can be done to trigger the errors.

### Payload Version

2.14.1

### Adapters and Plugins

"@payloadcms/bundler-webpack": "^1.0.6", "@payloadcms/db-mongodb": "^1.5.1", "@payloadcms/plugin-cloud-storage": "^1.1.2", "@payloadcms/plugin-nested-docs": "^1.0.12", "@payloadcms/plugin-seo": "^2.3.1", "@payloadcms/richtext-lexical": "^0.9.1", "@aws-sdk/client-s3": "^3.556.0", "@aws-sdk/lib-storage": "^3.556.0", "aws-crt": "^1.21.2"

> I found a thread that deals with this specific issue: [aws/aws-sdk-js-v3#3560](https://github.com/aws/aws-sdk-js-v3/issues/3560)
> 
> I would give a few of these solutions a try.

Hello, Hi can I implement this solution using the adapter? 
```ts requestHandler: new NodeHttpHandler({
    httpsAgent: new Agent({
      maxSockets: 500,

      // keepAlive is a default from AWS SDK. We want to preserve this for
      // performance reasons.
      keepAlive: true,
      keepAliveMsecs: 1000,
    }),
    socketTimeout: 5000,
  }),
``` 
When I tried to  import { NodeHttpHandler } from '@smithy/node-http-handler'; 
I had this error
```Module not found: Error: Can't resolve 'http2' in '/home/node/app/node_modules/@smithy/node-http-handler/dist-es'```

> > I found a thread that deals with this specific issue: [aws/aws-sdk-js-v3#3560](https://github.com/aws/aws-sdk-js-v3/issues/3560)
> > I would give a few of these solutions a try.
> 
> Hello, Hi can I implement this solution using the adapter?
> 
> ```ts
>     httpsAgent: new Agent({
>       maxSockets: 500,
> 
>       // keepAlive is a default from AWS SDK. We want to preserve this for
>       // performance reasons.
>       keepAlive: true,
>       keepAliveMsecs: 1000,
>     }),
>     socketTimeout: 5000,
>   }),
> ```
> 
> When I tried to import { NodeHttpHandler } from '@smithy/node-http-handler'; I had this error `Module not found: Error: Can't resolve 'http2' in '/home/node/app/node_modules/@smithy/node-http-handler/dist-es'`

My hosting provider gave me a similar solution to try

```
import https from "https"; 

              requestHandler: {
                connectionTimeout: 10 * 1000,
                httpsAgent: new Agent({
                  maxSockets: Infinity,
                  keepAlive: false,
                }),
                keepAlive: false,
                maxSockets: Infinity,
              }
```

However, I got stuck as the same issue you were facing. I couldn't resolve the package within Payload.

> > > I found a thread that deals with this specific issue: [aws/aws-sdk-js-v3#3560](https://github.com/aws/aws-sdk-js-v3/issues/3560)
> > > I would give a few of these solutions a try.
> > 
> > 
> > Hello, Hi can I implement this solution using the adapter?
> > ```ts
> >     httpsAgent: new Agent({
> >       maxSockets: 500,
> > 
> >       // keepAlive is a default from AWS SDK. We want to preserve this for
> >       // performance reasons.
> >       keepAlive: true,
> >       keepAliveMsecs: 1000,
> >     }),
> >     socketTimeout: 5000,
> >   }),
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > When I tried to import { NodeHttpHandler } from '@smithy/node-http-handler'; I had this error `Module not found: Error: Can't resolve 'http2' in '/home/node/app/node_modules/@smithy/node-http-handler/dist-es'`
> 
> My hosting provider gave me a similar solution to try
> 
> ```
> import https from "https"; 
> 
>               requestHandler: {
>                 connectionTimeout: 10 * 1000,
>                 httpsAgent: new Agent({
>                   maxSockets: Infinity,
>                   keepAlive: false,
>                 }),
>                 keepAlive: false,
>                 maxSockets: Infinity,
>               }
> ```
> 
> However, I got stuck on the same issue you are facing. I couldn't resolve the package within Payload.

@denolfe  could you help with this, please?

Unfortunately it doesn't work for me.

ERROR in ./node_modules/@smithy/node-http-handler/dist-es/node-http2-connection-manager.js 1:0-26
Module not found: Error: Can't resolve 'http2' in '/home/node/app/node_modules/@smithy/node-http-handler/dist-es'

payload config:

```
webpack: (config) => {
    return {
        ...config,
        resolve: {
            ...config.resolve,
            alias: {
                ...(config.resolve?.alias ? config.resolve.alias : {}),
                jsonwebtoken: mockModulePath,
            },
            fallback: {
                ...(config.resolve?.fallback ? config.resolve.fallback : {}),
                http: require.resolve("https-browserify"),
                http2: false,
                https: require.resolve("https-browserify"),
                url: require.resolve("url"),
            },
        },
    };
},
```

cloud storage adapter config:

```
const cloudStorageAdapter = s3Adapter({
    config: {
        region: process.env.AWS_REGION,
        credentials: {
            accessKeyId: process.env.AWS_ACCESS_KEY_ID!,
            secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY!,
        },
        requestHandler: new NodeHttpHandler({
            connectionTimeout: 5000,
            socketTimeout: 120000,
            httpAgent: new HttpAgent({ maxSockets: 500, keepAlive: true }),
            httpsAgent: new HttpsAgent({ maxSockets: 500, keepAlive: true }),
        }),
    },
    bucket: process.env.AWS_S3_BUCKET_NAME!,
});
```

--- a/packages/storage-s3/src/staticHandler.ts
+++ b/packages/storage-s3/src/staticHandler.ts
@@ -39,9 +39,11 @@ export const getHandler = ({ bucket, collection, getStorageClient }: Args): Stat
     try {
       const prefix = await getFilePrefix({ collection, filename, req })
 
+      const key = path.posix.join(prefix, filename)
+
       const object = await getStorageClient().getObject({
         Bucket: bucket,
-        Key: path.posix.join(prefix, filename),
+        Key: key,
       })
 
       if (!object.Body) {
@@ -74,7 +76,11 @@ export const getHandler = ({ bucket, collection, getStorageClient }: Args): Stat
       if (object.Body && isNodeReadableStream(object.Body)) {
         const stream = object.Body
         stream.on('error', (err) => {
-          console.error(err)
+          req.payload.logger.error({
+            err,
+            key,
+            msg: 'Error streaming S3 object, destroying stream',
+          })
           stream.destroy()
         })
       }

--- a/packages/storage-s3/src/staticHandler.ts
+++ b/packages/storage-s3/src/staticHandler.ts
@@ -1,6 +1,7 @@
 import type * as AWS from '@aws-sdk/client-s3'
 import type { StaticHandler } from '@payloadcms/plugin-cloud-storage/types'
 import type { CollectionConfig } from 'payload'
+import type { Readable } from 'stream'
 
 import { getFilePrefix } from '@payloadcms/plugin-cloud-storage/utilities'
 import path from 'path'
@@ -9,6 +10,18 @@ interface Args {
   bucket: string
   collection: CollectionConfig
   getStorageClient: () => AWS.S3
+}
+
+// Type guard for NodeJS.Readable streams
+const isNodeReadableStream = (body: unknown): body is Readable => {
+  return (
+    typeof body === 'object' &&
+    body !== null &&
+    'pipe' in body &&
+    typeof (body as any).pipe === 'function' &&
+    'destroy' in body &&
+    typeof (body as any).destroy === 'function'
+  )
 }
 
 // Convert a stream into a promise that resolves with a Buffer
@@ -39,7 +52,7 @@ export const getHandler = ({ bucket, collection, getStorageClient }: Args): Stat
       const objectEtag = object.ETag
 
       if (etagFromHeaders && etagFromHeaders === objectEtag) {
-        return new Response(null, {
+        const response = new Response(null, {
           headers: new Headers({
             'Accept-Ranges': String(object.AcceptRanges),
             'Content-Length': String(object.ContentLength),
@@ -47,6 +60,22 @@ export const getHandler = ({ bucket, collection, getStorageClient }: Args): Stat
             ETag: String(object.ETag),
           }),
           status: 304,
+        })
+
+        // Manually destroy stream before returning cached results to close socket
+        if (object.Body && isNodeReadableStream(object.Body)) {
+          object.Body.destroy()
+        }
+
+        return response
+      }
+
+      // On error, manually destroy stream to close socket
+      if (object.Body && isNodeReadableStream(object.Body)) {
+        const stream = object.Body
+        stream.on('error', (err) => {
+          console.error(err)
+          stream.destroy()
         })
       }
 

